### PR TITLE
Apim 1214 get plans -- rest api v2

### PR DESCRIPTION
**Pull request by jourdiw**
## Issue

https://gravitee.atlassian.net/browse/APIM-1214

## Description

- add polymorphism to plans --> V2 vs V4
- implement GET for listing plans

TODO: Post for creating a plan

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-vcizncswoc.chromatic.com)

### CODE REVIEW

**Review by loriepisicchio — APPROVED** [review]

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml [hunk 1]
```diff
@@ -486,7 +486,8 @@ paths:
                 - API Plans
             summary: Get the API's plans
             description: |
-                # FIXME add a description
+                List plans for a given API, always filtered by a `status` and the possibility to filter by `security`.
+                User must have API_PLAN[READ] and API_LOG[READ] permissions to access endpoint.
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml]: `API_PLAN[READ] and API_LOG[READ]`
or
`API_PLAN[READ] or API_LOG[READ]`
?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml [hunk 2]
```diff
@@ -2817,6 +2837,23 @@ components:
...
+                          items:
+                              $ref: "#/components/schemas/FlowV2"
+                      paths:
+                          type: object
+                          additionalProperties:
+                              type: array
+                              items:
+                                  $ref: "#/components/schemas/Rule"
+                      securityDefinition:
+                          type: string
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml]: I would put `object` like for other configuration-like fields. WDYT ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource.java [hunk 3]
```diff
@@ -94,22 +93,30 @@ public List<Plan> getApiPlans(
 
         GenericApiEntity genericApiEntity = apiSearchService.findGenericById(executionContext, apiId);
 
-        List<PlanEntity> entities = planService
+        List<GenericPlanEntity> plans = planSearchService
             .findByApi(executionContext, apiId)
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource.java]: I think we could use the `search` method of planSearchService. Maybe the plan status is not part of the `PlanQuery` but it should not be too hard to add it.
WDYT?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource.java [hunk 4]
```diff
@@ -94,22 +93,30 @@ public List<Plan> getApiPlans(
...
+            .filter(plan ->
+                planSecurityTypeParamList.isEmpty() ||
+                planSecurityTypeParamList.contains(PlanSecurityType.valueOf(plan.getPlanSecurity().getType()))
+            )
+            .sorted(comparingInt(GenericPlanEntity::getOrder))
             .map(this::filterSensitiveData)
             .collect(Collectors.toList());
 
-        return PlanMapper.INSTANCE.convertList(entities);
+        List<GenericPlanEntity> paginationData = computePaginationData(plans, paginationParam);
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource.java:109]: Besides the fact that in general, an API doesn't have too many plans, would it be a lot to do to implement a paginated search of plans?

**Review by phiz71 — COMMENTED** [review]
I have some questions

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml (hunk 1 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml]: To match the permissions in the past, it is an `and`. Need both rights.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml (hunk 2 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml]: It's now mapped to `security` in `BasePlan`, so I'm deleting the `securityDefinition` attribute.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml (hunk 1 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/resources/management-openapi-v2.yaml]: Will correct to -> `or` in code and doc

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    @Test
+    public void should_not_return_anything_if_no_api_id_specified() throws TechnicalException {
+        List<GenericPlanEntity> plans = planSearchService.search(
+            GraviteeContext.getExecutionContext(),
+            PlanQuery.builder().build(),
+            USER,
+            true
+        );
+
+        assert (plans.isEmpty());
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java]:
```suggestion
        assertNotNull(plans);
        assertEquals(0, plans.size());
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        Plan plan3 = createPlan("plan-3");
+        Plan plan4 = createPlan("plan-4");
+        Plan plan5 = createPlan("plan-5");
+        when(planRepository.findByApi(API_ID)).thenReturn(Set.of(plan1, plan2, plan3, plan4, plan5));
+
+        var rule = new Rule();
+        rule.setMethods(Set.of(HttpMethod.GET));
+        rule.setDescription("description of a rule");
+        rule.setEnabled(true);
+        var rules = List.of(rule);
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java:144]: Do you really need to set rules in the plan?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource_GetPlansTest.java [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+            .securityType(List.of(PlanSecurityType.JWT))
+            .status(List.of(PlanStatus.DEPRECATED))
+            .build();
+
+        var rule = new Rule();
+        rule.setMethods(Set.of(HttpMethod.GET));
+        rule.setDescription("description of a rule");
+        rule.setEnabled(true);
+
+        var rules = List.of(rule);
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource_GetPlansTest.java:205]: Do you really need to set rules in the plan?

**Review by phiz71 — COMMENTED** [review]
I have other questions 😅

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java (hunk 6 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java:144]: It was to see that the mapping didn't throw any errors

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java (hunk 5 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/v4/impl/PlanSearchServiceImpl_SearchTest.java]: Rebased

**Review by phiz71 — APPROVED** [review]

**Pull request merged**